src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py

```diff
@@ -0,0 +1,326 @@
+                        BinaryOperation.Operator.LBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    ubound = BinaryOperation.create(
+                        BinaryOperation.Operator.UBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    loop_bounds.append((lbound, ubound))
+                elif isinstance(shape, ArrayType.ArrayBounds):
+                    # array extent is defined in the array declaration
+                    loop_bounds.append(shape)
+                #  Note, the validate method guarantees that an else
+                #  clause is not required.
+        else:
+            # The validate method guarantees that this is an array
+            # reference.
+            loop_bounds = []
+            ndims = len(array_ref.children)
+            for shape in array_ref.children:
+                loop_bounds.append((shape.start.copy(), shape.stop.copy()))
+
+        # Determine the datatype of the array's values and create a
+        # scalar of that type
+        array_intrinsic = array_ref.symbol.datatype.intrinsic
+        array_precision = array_ref.symbol.datatype.precision
+        scalar_type = ScalarType(array_intrinsic, array_precision)
+
+        symbol_table = node.scope.symbol_table
+        assignment = node.ancestor(Assignment)
+
+        datatype = scalar_type
+        array_reduction = False
+        if dimension_ref and ndims > 1:
+            array_reduction = True
+            # We are reducing from one array to another
+            shape = []
+            for idx, bounds in enumerate(loop_bounds):
+                if int(dimension_literal.value)-1 == idx:
+                    pass
+                else:
+                    shape.append(bounds)
+            datatype = ArrayType(scalar_type, shape)
+
+        array_ref = node.children[0].detach()
+
+        # Create temporary variable based on the name of the intrinsic.
+        var_symbol = symbol_table.new_symbol(
+            f"{self._INTRINSIC_NAME.lower()}_var", symbol_type=DataSymbol,
+            datatype=datatype)
+        # Replace operation with a temporary variable.
+        if array_reduction:
+            array_indices = []
+            for idx in range(ndims-1):
+                array_indices.append(":")
+            reference = ArrayReference.create(var_symbol, array_indices)
+        else:
+            reference = Reference(var_symbol)
+        node.replace_with(reference)
+
+        # Create the loop iterators
+        loop_iterators = []
+        array_iterators = []
+        for idx in range(ndims):
+            loop_iterator = symbol_table.new_symbol(
+                f"i_{idx}", symbol_type=DataSymbol, datatype=INTEGER_TYPE)
+            loop_iterators.append(loop_iterator)
+            if array_reduction and idx != int(dimension_literal.value)-1:
```

Comment:
Again, I think that if `dim` is a `Reference` (but not a constant value) we could get to here with `dimension_literal` still being None?

Reply:
We could, but thanks to your earlier comment this is now picked up in the validation.